File: arbitrum/std/queue.py
# Copyright 2019, Offchain Labs, Inc.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from ..annotation import modifies_stack
from . import make_boundedq_type
from .. import value
from .struct import Struct


# a queue is just a boundedq; when the boundedq gets full
# we generate a new, bigger one

def make_queue_type(typ):
    boundedq = make_boundedq_type(typ)
    queue_type = Struct(f"queue[{typ}]", [
        ("boundedq", boundedq.typ)
    ])
    # stack_type.fields[]

    class Queue:
        @staticmethod
        def make():
            return boundedq.make(8)

        @staticmethod
        @modifies_stack([], [queue_type.typ])
        def new(vm):
            vm.push(Queue.make())
            vm.cast(queue_type.typ)

        @staticmethod
        @modifies_stack([queue_type.typ], [value.IntType()])
        def isempty(vm):
            # q -> isempty
            queue_type.get("boundedq")(vm)
            boundedq.isempty(vm)

        @staticmethod
        @modifies_stack([queue_type.typ, typ], [queue_type.typ])
        def put(vm):
            # q item -> updatedq
            queue_type.get("boundedq")(vm)
            vm.dup0()
            boundedq.isfull(vm)
            # bqisfull bq item
            vm.ifelse(lambda vm: [
                # bq item
                vm.dup0(),
                boundedq.struct.get("nextPutIndex")(vm),
                vm.push(2),
                vm.mul(),
                boundedq.new(vm),
                vm.swap1(),
                # oldbq newbq item
                vm.while_loop(lambda vm: [
                    vm.dup0(),
                    boundedq.isempty(vm),
                    vm.push(0),
                    vm.eq()
                    # (oldbq is nonempty) oldbq newbq item
                ], lambda vm: [
                    # oldbq newbq item
                    boundedq.get(vm),
                    # moveitem oldbq newbq item
                    vm.swap1(),
                    vm.swap2(),
                    # newbq moveitem oldbq item
                    boundedq.put(vm),
                    # newbq oldbq item
                    vm.swap1(),
                ]),
                # oldbq newbq item
                vm.pop()
            ])
            # bq item
            boundedq.put(vm)
            queue_type.set_val("boundedq")(vm)

        @staticmethod
        @modifies_stack([queue_type.typ], [typ, queue_type.typ])
        def get(vm):
            # assume queue is non-empty
            # q -> item q
            queue_type.get("boundedq")(vm)
            boundedq.get(vm)
            vm.swap1()
            queue_type.set_val("boundedq")(vm)
            vm.swap1()
    Queue.typ = queue_type.typ
    return Queue

queue = make_queue_type(value.ValueType())
queue_tup = make_queue_type(value.TupleType())

File: packages/arb-tx-aggregator/batcher/queue.go
/*
 * Copyright 2020, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package batcher

import (
	"container/heap"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/pkg/errors"
	"math/rand"
)

// An TxHeap is a min-heap of transactions sorted by nonce.
type TxHeap []*types.Transaction

func (h TxHeap) Len() int           { return len(h) }
func (h TxHeap) Less(i, j int) bool { return h[i].Nonce() < h[j].Nonce() }
func (h TxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *TxHeap) Push(x interface{}) {
	*h = append(*h, x.(*types.Transaction))
}

func (h *TxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

type txQueue struct {
	txes        TxHeap
	txesByNonce map[uint64]*types.Transaction
	maxNonce    uint64
}

func newTxQueue() *txQueue {
	return &txQueue{
		txes:        nil,
		txesByNonce: make(map[uint64]*types.Transaction),
		maxNonce:    0,
	}
}

func (q *txQueue) addTransaction(tx *types.Transaction) error {
	if _, ok := q.txesByNonce[tx.Nonce()]; ok {
		return errors.New("transaction replacement not supported")
	}

	q.txesByNonce[tx.Nonce()] = tx
	heap.Push(&q.txes, tx)

	if tx.Nonce() > q.maxNonce {
		q.maxNonce = tx.Nonce()
	}
	return nil
}

func (q *txQueue) Empty() bool {
	return len(q.txes) == 0
}

func (q *txQueue) Peek() *types.Transaction {
	if q == nil || len(q.txes) == 0 {
		return nil
	}
	return q.txes[0]
}

func (q *txQueue) Pop() *types.Transaction {
	tx := heap.Pop(&q.txes).(*types.Transaction)
	delete(q.txesByNonce, tx.Nonce())
	return tx
}

type txQueues struct {
	queues   map[common.Address]*txQueue
	accounts []common.Address
}

func newTxQueues() *txQueues {
	return &txQueues{
		queues:   make(map[common.Address]*txQueue),
		accounts: nil,
	}
}

func (q *txQueues) addTransaction(tx *types.Transaction, sender common.Address) error {
	queue, ok := q.queues[sender]
	if !ok {
		queue = newTxQueue()
		q.queues[sender] = queue
		q.accounts = append(q.accounts, sender)
	}
	return queue.addTransaction(tx)
}

func (q *txQueues) removeTxFromAccountAtIndex(i int) {
	q.queues[q.accounts[i]].Pop()
}

func (q *txQueues) maybeRemoveAccountAtIndex(i int) {
	account := q.accounts[i]
	if q.queues[account].Empty() {
		delete(q.queues, account)
		q.accounts[i] = q.accounts[len(q.accounts)-1]
		q.accounts = q.accounts[:len(q.accounts)-1]
	}
}

func popRandomTx(b batch, queuedTxes *txQueues) (*types.Transaction, int, bool) {
	queuedCount := int32(len(queuedTxes.accounts))
	if queuedCount == 0 {
		return nil, 0, false
	}
	index := int(rand.Int31n(queuedCount))
	first := true
	lastIndex := index
	index--
	for {
		if len(queuedTxes.accounts) == 0 {
			return nil, 0, false
		}
		index++
		if index == len(queuedTxes.accounts) {
			index = 0
		}
		if !first && index == lastIndex {
			return nil, 0, false
		}

		first = false
		account := queuedTxes.accounts[index]
		nextAccount := queuedTxes.queues[account]
		tx := nextAccount.Peek()
		// No tx in account
		if tx == nil {
			queuedTxes.maybeRemoveAccountAtIndex(index)
			continue
		}

		// err param can be ignored
		action, _ := b.validateTx(tx)
		switch action {
		case REMOVE:
			queuedTxes.removeTxFromAccountAtIndex(index)
		case SKIP:
		case FULL:
			return nil, 0, true
		case ACCEPT:
			queuedTxes.removeTxFromAccountAtIndex(index)
			return tx, index, true
		}
	}
}
